### PR TITLE
GUARD-3019 Ensure .NET Standard Compatibility

### PR DESCRIPTION
**Pull request by maxim-skuvault**
[GUARD-3019]
Summary: Upgrade the included NuGet packages if they aren't compatible with .NET Standard

#### Background
In a prior ticket [GUARD-2792] converted this project to .NET Standard 2.0/.NET Framework 4.8 multi-target, in order to meet the upcoming cloud requirement. However, as result of this some NuGet packages used in this solution showed a warning that they don't have a .NET Standard version. They were restored as .NET framework, which will not work in environments once we migrate to .NET Standard.

#### About these changes
- Replace CuttingEdge.Conditions with SkuVault.Integration.Core.Helpers.ValidationHelper constructor (and one method) validation:
  - Update parameter validation code to use ValidationHelper
  - Since ValidationHelper runs after the params are assigned to class properties, reworked the StoreUrl property in [ThreeDCartConfig.cs](https://github.com/skuvault-integrations/3dCartAccess/pull/10/files#diff-401d82d64baeb251c8b51dd27801124a9ee9d2bcd92a78e036a668a571a301f2) & [RestThreeDCartConfig.cs](https://github.com/skuvault-integrations/3dCartAccess/pull/10/files#diff-bcf00f9318d57415f82c76aed5a16903a863555d2a395f45ac390519c4fc226c). Ensured that it still works the same via pre-refactor unit tests
  - Add tests to verify the refactor. Ran them before the refactor first
  - Add skuvault-integrations package source, to reference SkuVault.Integration.Core package
  - Upgrade the NUnit package since the earlier didn't support some method we commonly use
  - Remove any reference to CuttingEdge.Conditions in the solution
-  Upgrade Netco package to 3.0.0 to be compatible with .NET Standard 2.0. This automatically removed an implicit dependency of this package on CuttingEdge.Conditions
- Upgrade ServiceStack.Text to 6.2.0 (compatible with .NET Standard 2.0). We use this version in other SkuVault solutions. After this all .NET Standard-related build warnings are gone!
- [Bug Fix] Since with "package on build" we always just package as Debug (even for release package) changed this to retry 10 times, so that we don't accidentally reduce retries in prod to just 1. This was a side-effect of earlier .NET Standard ticket https://github.com/skuvault-integrations/3dCartAccess/pull/4.
- [Refactor] Move integration tests into the "Integration" folder and mark them as `[ Explicit ]` as is our practice
- Reference the System.Runtime.CompilerServices.Unsafe package in the test project. Was getting an error about inconsistent versions when running many integration tests, due to an implicit reference in 3dCartAccess project. Probably, because "System.Runtime.CompilerServices.InternalsVisibleToAttribute" was added.

# PR Readiness Checklist

- [x] I have updated all relevant configuration
- [x] Followed [development conventions][1] to the best of my ability
- [x] Code is documented, particularly public interfaces and hard-to-understand areas
- [x] Tests are updated / added and all pass
- [x] Performed a self-review of my own code
- [x] Acceptance criterias are confirmed by testing changes in UI (or another appropriate way)
- [ ] Confluence documentation is updated, if needed
- [x] New code does not generate new warnings

[1]: https://agileharbor.atlassian.net/wiki/spaces/DEV/pages/1114130/Conventions

[GUARD-3019]: https://agileharbor.atlassian.net/browse/GUARD-3019?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[GUARD-2792]: https://agileharbor.atlassian.net/browse/GUARD-2792?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by SShumer** [comment]
@maxim-skuvault Prpbably I was wrong but I expected you to migrate SOAP service references too. 
image
Otherwise we'll lose control over it and won't be able to update it if interface is changes. Something like that:
https://github.com/skuvault-integrations/3dCartAccess/pull/11

I'm not sure it works (I don't have tests credential files) but I believe it should be something like that.

If you think it's out of this ticket scope, don't we need to create a new ticket than?

And I don't understand why do we have 
`<Compile Include="<file_name>" /> ...`
items for all files in the project file. I thought it's not needed for the net standard app.

File: src/.nuget/NuGet.Config [hunk 1]
```diff
@@ -4,4 +4,10 @@
 	<solution>
 		<add key="disableSourceControlIntegration" value="true" />
 	</solution>
+  <packageSources>
```
[the remark this answers is not in the document]
Comment (maxim-skuvault) [reply src/.nuget/NuGet.Config]: @SShumer Fixed. Rider has added this but might as well ensure the entire file is formatted as tabs

File: src/ThreeDCartAccess/RestApi/Models/Configuration/RestThreeDCartConfig.cs [hunk 2]
```diff
@@ -1,32 +1,61 @@
...
 
 namespace ThreeDCartAccess.RestApi.Models.Configuration
 {
 	public sealed class RestThreeDCartConfig
 	{
 		public string BaseUrl => "http://apirest.3dcart.com/3dCartWebAPI/v2";
 
-		public string StoreUrl{ get; }
+		private readonly string storeUrlRaw;
+		public string StoreUrl => this.storeUrlRaw.ToLower().TrimEnd( '\\', '/' ).Replace( "https://", "" ).Replace( "http://", "" ).Replace( "www.", "" );
```
[the remark this answers is not in the document]
Comment (maxim-skuvault) [reply src/ThreeDCartAccess/RestApi/Models/Configuration/RestThreeDCartConfig.cs]: @SShumer I like this idea. As a bonus, validation will also fail if the raw storeUrl becomes "" after the "standardization" here. For example, raw "https://www.\\" => "". If we only validate that the raw storeUrl is null or empty then we won't catch this on validation and we'll try to call the API with an invalid storeUrl. Committed.

File: src/ThreeDCartAccess/SoapApi/Models/Configuration/ThreeDCartConfig.cs [hunk 3]
```diff
@@ -1,24 +1,41 @@
...
+﻿using System.Collections.Generic;
+using SkuVault.Integrations.Core.Helpers;
 
 namespace ThreeDCartAccess.SoapApi.Models.Configuration
 {
 	public sealed class ThreeDCartConfig
 	{
-		public string StoreUrl{ get; private set; }
+		private readonly string storeUrlRaw;
+		public string StoreUrl => this.storeUrlRaw.ToLower().TrimEnd( '\\', '/' ).Replace( "https://", "" ).Replace( "http://", "" ).Replace( "www.", "" );
```
[the remark this answers is not in the document]
Comment (maxim-skuvault) [reply src/ThreeDCartAccess/SoapApi/Models/Configuration/ThreeDCartConfig.cs]: @SShumer [Done](https://github.com/skuvault-integrations/3dCartAccess/pull/10/files#r1263798028)

**Comment by maxim-skuvault** [comment]
> @maxim-skuvault Prpbably I was wrong but I expected you to migrate SOAP service references too. image Otherwise we'll lose control over it and won't be able to update it if interface is changes. Something like that: #11
> 
> I'm not sure it works (I don't have tests credential files) but I believe it should be something like that.
> 
> If you think it's out of this ticket's scope, don't we need to create a new ticket than?
> 
> And I don't understand why do we have `<Compile Include="<file_name>" /> ...` items for all files in the project file. I thought it's not needed for the net standard app.

@SShumer In this ticket we're just replacing / removing packages not compatible with .NET Standard. I'm still keeping it .NET Standard / Framework multi-target since for the deadline Core said this will be sufficient. We can try to convert the project to just .NET Standard in another ticket. When I tried this in the initial 3dCart .NET Standard migration ticket, I ran into a lot of issues with these service references and I think other issues (prob. left comments). I tried several ways to regenerate these files and there were issues. This integration is used by 12 tenants. As long as we meet the deprecation deadline (some time in 2024) I don't know if it'll be a priority to make this pure .NET Standard

**Review by SShumer — APPROVED** [review]

File: src/ThreeDCartAccess/RestApi/Models/Configuration/RestThreeDCartConfig.cs [hunk 4]
```diff
@@ -1,32 +1,64 @@
...
+			var validationErrors = new List<string>();
+			if ( string.IsNullOrWhiteSpace( this.StoreUrl ) )
+			{
+				validationErrors.Add( $"{nameof( this.StoreUrl )} is null or white space" );
+			}
+			if ( string.IsNullOrWhiteSpace( this.Token ) )
+			{
+				validationErrors.Add( $"{nameof( this.Token )} is null or white space" );
+			}
+			if ( this.TimeZone is < -12 or > 12 )
```
[the remark this answers is not in the document]
Comment (Victoria-SV) [reply src/ThreeDCartAccess/RestApi/Models/Configuration/RestThreeDCartConfig.cs:40]: resolving. Do you think it's worth creating a TD ticket to research this validation logic further and possibly remove it?

**Pull request merged**